File: classical-physics/mechanics/simple-pendulum/classic-simple-pendulum.py
# Libraries
import numpy as np
import matplotlib.pyplot as plt

# Plot parameters
plt.rcParams.update({
    'lines.linewidth': 2,     # linewidth
    'text.usetex': True,      # LaTeX font
    'font.family': 'serif',   # Serif family
    'font.size': 16,          # font size
    'axes.titlesize': 20,     # title size
    'axes.grid': True,        # grid
    'grid.linestyle': "-.",   # grid style
})

# Physical parameters
g = 9.81 # gravity (kg/m^2)
l = 1    # length  (m)

# Initial conditions
θ0 = np.pi/4   # initial angle            (rad)
ω0 = 3.0       # initial angular velocity (rad/s)

# Time parameters
t_0 = 0                       # initial time (s)
t_f = 10                      # final time   (s)
Δt = 0.01                     # step size    (s)
n = int((t_f - t_0) / Δt) + 1 # iterations
t = np.linspace(t_0, t_f, n)  # time array  (s)

# Fourth-Order Runge-Kutta function
def RK4(f, x0, y0, z0, h):
    k1y = h * f(x0, y0, z0)[0]
    k1z = h * f(x0, y0, z0)[1]
    #
    k2y = h * f(x0 + h / 2, y0 + k1y / 2, z0 + k1z / 2)[0]
    k2z = h * f(x0 + h / 2, y0 + k1y / 2, z0 + k1z / 2)[1]
    #
    k3y = h * f(x0 + h / 2, y0 + k2y / 2, z0 + k2z / 2)[0]
    k3z = h * f(x0 + h / 2, y0 + k2y / 2, z0 + k2z / 2)[1]
    #
    k4y = h * f(x0 + h, y0 + k3y, z0 + k3z)[0]
    k4z = h * f(x0 + h, y0 + k3y, z0 + k3z)[1]

    # Approximation
    y1 = float(y0 + (k1y + 2 * k2y + 2 * k3y + k4y) / 6)
    z1 = float(z0 + (k1z + 2 * k2z + 2 * k3z + k4z) / 6)
    return y1, z1

# Differential Equations
def f(t, θ, ω):
    return ω, -(g / l) * np.sin(θ)

# Initial arrays
θ, θ[0] = np.zeros(n), θ0
ω, ω[0] = np.zeros(n), ω0

# RK4 method evaluation
for i in range(n - 1):
    θ[i + 1], ω[i + 1] = RK4(f, t[i], θ[i], ω[i], Δt)

plt.figure(figsize = (10, 4.5))

# Plot
plt.subplot(1, 2, 1)
plt.plot(t, θ, 'b-')
plt.xlabel(r'$t$ (s)')
plt.ylabel(r'$\theta(t)$ [rad]')
plt.title(r'Angle')

plt.subplot(1, 2, 2)
plt.plot(t, ω, 'r-')
plt.xlabel(r'$t$ (s)')
plt.ylabel(r'$\omega(t)$ [rad/s]')
plt.title(r'Angular velocity')
plt.tight_layout()
plt.gcf()
plt.savefig("pyplot2.pdf")

File: classical-physics/mechanics/simple-pendulum/classic_simple_pendulum.jl
# Libraries
using LinearAlgebra
using Plots, ColorSchemes, LaTeXStrings

# Physical parameters
g = 9.81 # gravity (kg/m^2)
l = 1    # length  (m)

# Initial conditions
θ₀ = π/4 # initial angle   (rad)
ω₀ = 3.0 # initial angular (rad/s)

# Time parameters
t₀ = 0                        # initial time (s)
t₁ = 10                       # final time   (s)
Δt = 0.01                     # step size    (s)
n = Int64((t₁ - t₀) / Δt)     # iterations
t = t₀:Δt:(t₁ - Δt)           # time array   (s)

# Fourth-Order Runge-Kutta function
function RK4(f, x₀, y₀, z₀, h)
    k1y = h .* f(x₀, y₀, z₀)[1]
    k1z = h .* f(x₀, y₀, z₀)[2]
    #
    k2y = h .* f(x₀ + h / 2, y₀ + k1y / 2, z₀ + k1z / 2)[1]
    k2z = h .* f(x₀ + h / 2, y₀ + k1y / 2, z₀ + k1z / 2)[2]
    #
    k3y = h .* f(x₀ + h / 2, y₀ + k2y / 2, z₀ + k2z / 2)[1]
    k3z = h .* f(x₀ + h / 2, y₀ + k2y / 2, z₀ + k2z / 2)[2]
    #
    k4y = h .* f(x₀ + h, y₀ + k3y, z₀ + k3z)[1]
    k4z = h .* f(x₀ + h, y₀ + k3y, z₀ + k3z)[2]

    # Approximation
    y1 = Float64(y₀ + (k1y + 2 * k2y + 2 * k3y + k4y) / 6)
    z1 = Float64(z₀ + (k1z + 2 * k2z + 2 * k3z + k4z) / 6)
    return y1, z1
end

# Differential equations
function f(t, θ, ω)
    return ω, -(g / l) .* sin.(θ)
end

# Initial arrays
θ, θ[1] = zeros(n), θ₀
ω, ω[1] = zeros(n), ω₀

# RK4 method evaluation
for i in 1:(n - 1)
    θ[i + 1], ω[i + 1] = RK4(f, t[i], θ[i], ω[i], Δt)
end

# Plot
default(fontfamily = "Computer modern", legend = false, tickfont = font(12), guidefont = font(20), titlefont = font(24), linewidth = 2)
p1 = plot(t, θ, title = L"\mathrm{Angle}", xlabel = L"$t$", ylabel = L"$\theta(t)\:[\mathrm{rad}]$", color = "blue")
p2 = plot(t, ω, title = L"\mathrm{Angular}\:\: \mathrm{velocity}", xlabel = L"$t$", ylabel = L"$\omega(t)\:[\mathrm{rad}/\mathrm{s}]$", color = "orange")
plot(p1, p2, layout = (1, 2), size = (1200, 400), margin = 10Plots.mm)
savefig("juliaplot2.pdf")